File: apps/node-api/src/scoring.ts
import { PrismaClient } from '@prisma/client';

const prisma = new PrismaClient();

export interface CarbonScoreResponse {
	productId: string;
	co2eKg: number;
	tier: 'Low' | 'Medium' | 'High';
	confidenceScore: 'low' | 'medium' | 'high';
	calculationMethod: string;
	comparisonStatement: string;
	breakdown: {
		productMaterials: number;
		packaging: number;
		transportEstimate: number;
	};
	categoryId?: string;
}

export interface OffsetOption {
	id: string;
	name: string;
	type: string;
	costPerTonCents: number;
	calculatedCostCents: number;
}

// In-memory fallback dataset for mock testing
export const mockCatalog = [
	{
		sku: 'sku-almondmilk-002',
		title: 'Almond Milk 1L',
		categoryId: 'category-dairy-eggs',
		weightGrams: 1000,
		gtin: '000000000002',
	},
	{
		sku: 'sku-beefsteak-003',
		title: 'Premium Beef Steak',
		categoryId: 'category-meat-seafood',
		weightGrams: 400,
		gtin: '000000000003',
	},
	{
		sku: 'sku-oatmilk-local',
		title: 'Oat Milk 1L (Local)',
		categoryId: 'category-dairy-eggs',
		weightGrams: 1000,
		gtin: '000000000004',
	},
];

export const mockFactors: Record<string, { co2ePerKg: number; sourceCode: string }> = {
	'category-dairy-eggs': { co2ePerKg: 2.50, sourceCode: 'AGRI_DAIRY' },
	'category-meat-seafood': { co2ePerKg: 12.00, sourceCode: 'AGRI_MEAT' },
};

// Map kg CO2e to Low/Medium/High thresholds
export function mapScoreToTier(co2e: number): 'Low' | 'Medium' | 'High' {
	if (co2e < 1.0) {
		return 'Low';
	} else if (co2e >= 5.0) {
		return 'High';
	}
	return 'Medium';
}

// Check database connectivity
async function isDbOnline(): Promise<boolean> {
	if (!process.env.DATABASE_URL) return false;
	try {
		await prisma.$queryRaw`SELECT 1`;
		return true;
	} catch (e) {
		return false;
	}
}

// Main carbon scoring calculation engine
export async function resolveProductScore(
	partnerId: string,
	productId: string,
	weightGrams?: number,
	title?: string,
	categoryName?: string
): Promise<CarbonScoreResponse> {
	const dbActive = await isDbOnline();
	let co2ePerKg = 0;
	let confidence: 'low' | 'medium' | 'high' = 'low';
	let calcMethod = 'global_default_baseline';
	let categoryId = '';
	let finalWeightGrams = weightGrams || 0;

	// 1. Check direct SKU override
	if (dbActive) {
		try {
			const override = await prisma.partnerOverride.findFirst({
				where: { partnerId, sku: productId },
			});
			if (override) {
				co2ePerKg = Number(override.customFactor);
				confidence = 'high';
				calcMethod = 'exact_sku_match';
			}
		} catch (e) {
			// ignore and proceed
		}
	} else {
		// Mock override check
		if (productId === 'sku-bananas-001') {
			co2ePerKg = 0.35;
			confidence = 'high';
			calcMethod = 'exact_sku_match';
		}
	}

	if (co2ePerKg === 0) {
		// 2. Query Custom Product Catalog
		let catalogItem: any = null;

		if (dbActive) {
			try {
				catalogItem = await prisma.productCatalog.findFirst({
					where: {
						partnerId,
						OR: [
							{ sku: productId },
							{ gtin: productId },
							{ catalogItemId: productId },
						],
					},
				});
			} catch (e) {
				// ignore
			}
		} else {
			// Mock catalog lookup
			catalogItem = mockCatalog.find(
				(c) => c.sku === productId || c.gtin === productId
			) || null;
		}

		if (catalogItem) {
			categoryId = catalogItem.categoryId;
			if (finalWeightGrams <= 0 && catalogItem.weightGrams) {
				finalWeightGrams = catalogItem.weightGrams;
			} else if (finalWeightGrams <= 0 && catalogItem.attributes) {
				// Parse attributes JSON if available
				const attrs = catalogItem.attributes as any;
				if (attrs && typeof attrs.weight_grams === 'number') {
					finalWeightGrams = attrs.weight_grams;
				}
			}

			// Check category-level override
			if (dbActive) {
				try {
					const catOverride = await prisma.partnerOverride.findFirst({
						where: { partnerId, categoryId, sku: null },
					});
					if (catOverride) {
						co2ePerKg = Number(catOverride.customFactor);
						confidence = 'high';
						calcMethod = 'category_override';
					}
				} catch (e) {
					// ignore
				}
			}

			// Query category factor
			if (co2ePerKg === 0) {
				if (dbActive) {
					try {
						const factor = await prisma.emissionFactor.findFirst({
							where: {
								categoryId,
								validFrom: { lte: new Date() },
								OR: [
									{ validTo: null },
									{ validTo: { gte: new Date() } },
								],
							},
						});
						if (factor) {
							co2ePerKg = Number(factor.co2ePerKg);
							confidence = 'high';
							calcMethod = 'exact_sku_match';
						}
					} catch (e) {
						// ignore
					}
				} else {
					// Mock factors mapping
					const factor = mockFactors[categoryId];
					if (factor) {
						co2ePerKg = factor.co2ePerKg;
						confidence = 'high';
						calcMethod = 'exact_sku_match';
					}
				}
			}

			// Check parent category fallback
			if (co2ePerKg === 0 && dbActive) {
				try {
					const category = await prisma.category.findUnique({
						where: { categoryId },
						include: { parent: true },
					});
					if (category && category.parentId) {
						const parentFactor = await prisma.emissionFactor.findFirst({
							where: {
								categoryId: category.parentId,
								validFrom: { lte: new Date() },
								OR: [
									{ validTo: null },
									{ validTo: { gte: new Date() } },
								],
							},
						});
						if (parentFactor) {
							co2ePerKg = Number(parentFactor.co2ePerKg);
							confidence = 'medium';
							calcMethod = 'parent_category_fallback';
						}
					}
				} catch (e) {
					// ignore
				}
			}
		}

		// 3. Fallback on query params if not resolved
		if (co2ePerKg === 0 && categoryName) {
			if (dbActive) {
				try {
					const category = await prisma.category.findFirst({
						where: { name: { equals: categoryName, mode: 'insensitive' } },
					});
					if (category) {
						categoryId = category.categoryId;
						const factor = await prisma.emissionFactor.findFirst({
							where: { categoryId },
						});
						if (factor) {
							co2ePerKg = Number(factor.co2ePerKg);
							confidence = 'medium';
							calcMethod = 'category_fallback';
						}
					}
				} catch (e) {
					// ignore
				}
			} else {
				// Mock match on name fallback
				if (categoryName.toLowerCase() === 'dairy & eggs') {
					categoryId = 'category-dairy-eggs';
					co2ePerKg = 2.50;
					confidence = 'medium';
					calcMethod = 'category_fallback';
				}
			}
		}

		// 4. Default global baseline
		if (co2ePerKg === 0) {
			co2ePerKg = 1.5;
			confidence = 'low';
			calcMethod = 'global_default_baseline';
		}
	}

	const weightKg = finalWeightGrams > 0 ? finalWeightGrams / 1000.0 : 1.0;
	const co2eKg = Math.round(co2ePerKg * weightKg * 100) / 100;
	const tier = mapScoreToTier(co2eKg);

	// Generate comparisons
	const phoneCharges = Math.round(co2eKg / 0.0082);
	const comparisonStatement = `equivalent to charging ${phoneCharges} smartphones`;

	// Product breakdown
	const breakdown = {
		productMaterials: Math.round(co2eKg * 0.80 * 100) / 100,
		packaging: Math.round(co2eKg * 0.15 * 100) / 100,
		transportEstimate: Math.round(co2eKg * 0.05 * 100) / 100,
	};

	return {
		productId,
		co2eKg,
		tier,
		confidenceScore: confidence,
		calculationMethod: calcMethod,
		comparisonStatement,
		breakdown,
		categoryId,
	};
}

// Finds a substitute product in same category with lower footprint
export async function getAlternativesSuggestions(
	partnerId: string,
	categoryId: string,
	currentSku: string,
	currentScore: number
): Promise<any | null> {
	if (!categoryId) return null;
	const dbActive = await isDbOnline();

	let candidates: any[] = [];
	if (dbActive) {
		try {
			candidates = await prisma.productCatalog.findMany({
				where: {
					partnerId,
					categoryId,
					sku: { not: currentSku },
				},
				take: 5,
			});
		} catch (e) {
			// ignore
		}
	} else {
		// Mock suggestions candidates
		candidates = mockCatalog.filter(
			(c) => c.categoryId === categoryId && c.sku !== currentSku
		);
	}

	for (const alt of candidates) {
		const score = await resolveProductScore(
			partnerId,
			alt.sku,
			alt.weightGrams
		);
		if (score.co2eKg < currentScore) {
			return {
				sku: alt.sku,
				title: alt.title,
				co2eKg: score.co2eKg,
				co2eSavedKg: Math.round((currentScore - score.co2eKg) * 100) / 100,
				priceCents: alt.sku.includes('oat') ? 420 : 399,
				reasoning: 'Oat production uses significantly less land and water than almonds.',
			};
		}
	}

	return null;
}

// Calculate commute transport options footprint
export async function getCommuteOptions(
	distanceKm: number
): Promise<any[]> {
	const dbActive = await isDbOnline();
	const modes = [
		{ name: 'diesel_van', speedKmh: 45, costPerKm: 0.12 },
		{ name: 'ev_car', speedKmh: 50, costPerKm: 0.08 },
		{ name: 'public_transit', speedKmh: 35, costPerKm: 0.05 },
		{ name: 'bicycle', speedKmh: 18, costPerKm: 0.00 },
	];

	const options: any[] = [];
	for (const m of modes) {
		let factorPerKm = 0.0;
		if (dbActive) {
			try {
				const factor = await prisma.transportFactor.findFirst({
					where: { modeName: m.name },
				});
				if (factor) {
					factorPerKm = Number(factor.co2ePerKmPassenger);
				}
			} catch (e) {
				// ignore
			}
		} else {
			// Mock factors
			if (m.name === 'diesel_van') factorPerKm = 0.18;
			else if (m.name === 'ev_car') factorPerKm = 0.04;
			else if (m.name === 'public_transit') factorPerKm = 0.02;
		}

		const co2eKg = Math.round(factorPerKm * distanceKm * 100) / 100;
		const durationMinutes = Math.round((distanceKm / m.speedKmh) * 60);
		const costCents = Math.round(m.costPerKm * distanceKm * 100);

		options.push({
			mode: m.name,
			durationMinutes,
			costCents,
			co2eKg,
			tier: mapScoreToTier(co2eKg),
		});
	}

	return options;
}

File: apps/api/scoring.go
package main

import (
	"fmt"
	"math"
	"strings"
)

// ResolveProductScore calculates the carbon footprint based on the estimation tree
func ResolveProductScore(partnerID string, sku string, weightGrams int, title string, categoryName string) (CarbonScoreResponse, error) {
	var co2ePerKg float64
	var confidence = "low"
	var calcMethod = "global_default_baseline"
	var categoryID string

	// Determine initial weight
	weightKg := float64(weightGrams) / 1000.0
	if weightKg <= 0 {
		weightKg = 1.0 // Default fallback weight (1 kg)
	}

	// 1. Check Exact SKU Override
	skuOverride, foundOverride, err := GetSKUOverride(partnerID, sku)
	if err == nil && foundOverride {
		co2ePerKg = skuOverride
		confidence = "high"
		calcMethod = "exact_sku_match"
	} else {
		// 2. Check Custom Product Catalog
		catalogItem, err := GetCatalogItem(partnerID, sku)
		if err == nil && catalogItem != nil {
			categoryID = catalogItem.CategoryID
			// If request didn't supply weight, use weight from catalog attributes
			if weightGrams <= 0 && catalogItem.WeightGrams > 0 {
				weightKg = float64(catalogItem.WeightGrams) / 1000.0
			}

			// Check category-level override
			catOverride, foundCatOverride, err := GetCategoryOverride(partnerID, categoryID)
			if err == nil && foundCatOverride {
				co2ePerKg = catOverride
				confidence = "high"
				calcMethod = "category_override"
			} else {
				// Query category factor
				factor, err := GetCategoryFactor(categoryID)
				if err == nil && factor != nil {
					co2ePerKg = factor.CO2ePerKg
					confidence = "high"
					calcMethod = "exact_sku_match"
				} else {
					// Check parent category fallback
					parentID, err := GetParentCategory(categoryID)
					if err == nil && parentID != "" {
						parentFactor, err := GetCategoryFactor(parentID)
						if err == nil && parentFactor != nil {
							co2ePerKg = parentFactor.CO2ePerKg
							confidence = "medium"
							calcMethod = "parent_category_fallback"
						}
					}
				}
			}
		}

		// 3. Check Category Name query param if no catalog item or factor found yet
		if co2ePerKg == 0 && categoryName != "" {
			catIDByName, err := GetCategoryByName(categoryName)
			if err == nil && catIDByName != "" {
				categoryID = catIDByName
				factor, err := GetCategoryFactor(categoryID)
				if err == nil && factor != nil {
					co2ePerKg = factor.CO2ePerKg
					confidence = "medium"
					calcMethod = "category_fallback"
				}
			}
		}

		// 4. Keyword Text Classification System (if title/sku matches known words)
		if co2ePerKg == 0 && title != "" {
			inferredCategory := inferCategoryFromTitle(title)
			if inferredCategory != "" {
				catIDByName, err := GetCategoryByName(inferredCategory)
				if err == nil && catIDByName != "" {
					categoryID = catIDByName
					factor, err := GetCategoryFactor(categoryID)
					if err == nil && factor != nil {
						co2ePerKg = factor.CO2ePerKg
						confidence = "low"
						calcMethod = "text_classification_fallback"
					}
				}
			}
		}

		// 5. Global Default Baseline
		if co2ePerKg == 0 {
			co2ePerKg = 1.5 // Standard baseline: 1.5 kg CO2e/kg
			confidence = "low"
			calcMethod = "global_default_baseline"
		}
	}

	co2eTotal := co2ePerKg * weightKg

	// Determine Tier
	tier := "medium"
	if co2eTotal < 1.0 {
		tier = "low"
	} else if co2eTotal >= 5.0 {
		tier = "high"
	}

	// Generate comparisons
	// 1 smartphone charge is approx 0.0082 kg CO2e
	phoneCharges := math.Round(co2eTotal / 0.0082)
	comparison := fmt.Sprintf("equivalent to charging %.0f smartphones", phoneCharges)

	// Breakdown
	breakdown := map[string]float64{
		"product_materials":  math.Round(co2eTotal*0.80*100) / 100,
		"packaging":          math.Round(co2eTotal*0.15*100) / 100,
		"transport_estimate": math.Round(co2eTotal*0.05*100) / 100,
	}

	return CarbonScoreResponse{
		SKU:                 sku,
		CO2eKg:              math.Round(co2eTotal*100) / 100,
		ConfidenceScore:     confidence,
		Tier:                tier,
		CalculationMethod:   calcMethod,
		ComparisonStatement: comparison,
		Breakdown:           breakdown,
		CategoryID:          categoryID,
	}, nil
}

// inferCategoryFromTitle maps product names to seed categories
func inferCategoryFromTitle(title string) string {
	t := strings.ToLower(title)
	if strings.Contains(t, "banana") || strings.Contains(t, "apple") || strings.Contains(t, "fruit") || strings.Contains(t, "veg") {
		return "Fruits & Vegetables"
	}
	if strings.Contains(t, "milk") || strings.Contains(t, "cheese") || strings.Contains(t, "dairy") || strings.Contains(t, "yogurt") || strings.Contains(t, "egg") {
		return "Dairy & Eggs"
	}
	if strings.Contains(t, "beef") || strings.Contains(t, "chicken") || strings.Contains(t, "steak") || strings.Contains(t, "pork") || strings.Contains(t, "meat") || strings.Contains(t, "fish") || strings.Contains(t, "seafood") {
		return "Meat & Seafood"
	}
	if strings.Contains(t, "coffee") || strings.Contains(t, "tea") {
		return "Coffee & Tea"
	}
	if strings.Contains(t, "soda") || strings.Contains(t, "coke") || strings.Contains(t, "drink") || strings.Contains(t, "juice") || strings.Contains(t, "beverage") {
		return "Soft Drinks"
	}
	if strings.Contains(t, "phone") || strings.Contains(t, "iphone") || strings.Contains(t, "smartphone") || strings.Contains(t, "pixel") {
		return "Smartphones"
	}
	if strings.Contains(t, "laptop") || strings.Contains(t, "macbook") || strings.Contains(t, "computer") {
		return "Laptops"
	}
	return ""
}
